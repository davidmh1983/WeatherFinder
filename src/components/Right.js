import React, { useState } from "react";
//import "bootstrap/dist/css/bootstrap.min.css";
import { cityWeather } from "../api/fetchData";
import Results from "./Results";

const Right = () => {
  const [state, setState] = useState({ 
    information: [],
    error: undefined
  });

  const getWeather = async (e) => {
    e.preventDefault();
    const city = e.target.elements.city.value || "Madrid";
    const country = e.target.elements.country.value || "es";
    try{
      setState(structureData(await cityWeather(city, country), city, country));
    }catch(e){
      setState({error:'Please enter the values'});
    }
  }

  const structureData = (data, city, country) => {
    const placeChosen = city && country;
    if(data.message) {
      return ({
        information: [],
        error: data.message
      })
    }
    return (
      {
        information: [
          {
            check: placeChosen, label:'Temperature', result: placeChosen ? data.main.temp : undefined
          },
          {
            check: placeChosen && data.sys.country ,label: 'Location', result: placeChosen ? data.name +', '+ data.sys.country : undefined
          },
          {
            check:placeChosen, label: 'Humidity', result: placeChosen ? data.main.humidity : undefined
          },
          {
            check: placeChosen, label:'Description', result: placeChosen ? data.weather[0].description : undefined
          } 
        ],
        error: placeChosen ? "": "Please enter the values."
      }
    )
  }

  return(
    <div className="col-7 form-container">
      <form onSubmit={getWeather}>
        <input type="text" name="city" placeholder="Madrid" />
        <input type="text" name="country" placeholder="es" />
        <button>Get Weather</button>
      </form>
      <Results data={state}/>
    </div>
  )
}
export default Right;