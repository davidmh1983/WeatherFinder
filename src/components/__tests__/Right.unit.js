import React from 'react';
import Right from '../Right';
import { shallow } from '../../../enzyme';
import cityWeather from '../../api/fetchData';
import 'babel-polyfill';

jest.mock('../../api/fetchData');

const structuredData = (city, country) => {
  if (city === 'Madrid' && country === 'ES') {
    return {
      information: [
        {
          check: true, label: 'Temperature', result: 13.4,
        },
        {
          check: true, label: 'Location', result: 'Madrid, ES',
        },
        {
          check: true, label: 'Humidity', result: 45,
        },
        {
          check: true, label: 'Description', result: 'arida',
        },
      ],
      error: '',
    };
  }
  if ((city === '' && country === 'ES')
  || (city === 'Madrid' && country === '')
  || (city === '' && country === '')
  || (city === 'Madrid' && country === 'FR')) {
    return {
      information: [],
      error: 'city not found',
    };
  }
};

function checkTest(city, country, code) {
  const responseStructured = structuredData(city, country);
  const response = cityWeather(city, country, false);
  expect(response.cod).toEqual(code);
  if (code === '404') {
    expect(response.message).toEqual(responseStructured.error);
  }
  if (code === 200) {
    expect(response.name).toEqual(responseStructured.information[1].result.split(',')[0]);
    expect(response.main.temp).toEqual(responseStructured.information[0].result);
    expect(response.main.humidity).toEqual(responseStructured.information[2].result);
    expect(response.weather[0].description).toEqual(responseStructured.information[3].result);
  }
}

describe('Check response api', () => {
  it('Matches the snapshot', () => {
    const wrapper = shallow(<Right />);
    expect(wrapper).toMatchSnapshot();
  });

  it('city and country exist and city belong to that country', () => {
    checkTest('Madrid', 'ES', 200);
  });
  it('city exists and country doesnt exist or there is no country written', () => {
    checkTest('Madrid', '', '404');
  });
  it('city doesnt exist or there is no city written and country exists', () => {
    checkTest('', 'ES', '404');
  });
  it('city and country dont exist or there is nothing written', () => {
    checkTest('', '', '404');
  });
  it('city exists and country exists but city doesnt belong to that country', () => {
    checkTest('Madrid', 'FR', '404');
  });
});
