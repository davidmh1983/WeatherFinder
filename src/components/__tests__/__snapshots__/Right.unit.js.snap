// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check response api Matches the snapshot 1`] = `ShallowWrapper {}`;
